### ISSUE TITLE
Indexing doesn’t work after update to 3.5

### ISSUE THREAD

**Issue by matthias-ronge** [kitodo/kitodo-production#5625]
I did an update on a customers server from 3.4.3 to 3.5. I applied `migration_3-4-3_to_3-5-0.sql` without any errors. After the update, the index must be recreated. But indexing fails with the following stack trace in the log:

```
[ERROR] 2023-04-14 10:30:21.420 [pool-48-thread-1] SqlExceptionHelper - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'RECURSIVE process_children(id) as (        (SELECT id FROM process WHERE id = 28' at line 1
[ERROR] 2023-04-14 10:30:21.420 [pool-48-thread-1] IndexWorker - org.hibernate.exception.SQLGrammarException: could not extract ResultSet
javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1626) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.kitodo.data.database.persistence.TaskDAO.countTaskStatusForProcessAndItsAncestors(TaskDAO.java:183) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.converter.ProcessConverter.countTaskStatusOfProcess(ProcessConverter.java:166) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.converter.ProcessConverter.getTaskProgressPercentageOfProcess(ProcessConverter.java:248) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.type.ProcessType.convertProgressStatus(ProcessType.java:110) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.type.ProcessType.getJsonObject(ProcessType.java:80) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.type.ProcessType.getJsonObject(ProcessType.java:30) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.type.BaseType.createDocument(BaseType.java:53) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.type.BaseType.createDocuments(BaseType.java:60) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.data.elasticsearch.index.Indexer.performMultipleRequests(Indexer.java:113) ~[kitodo-data-management-3.5.0.jar:?]
        at org.kitodo.production.services.data.base.SearchService.addAllObjectsToIndex(SearchService.java:210) ~[classes/:3.5.0]
        at org.kitodo.production.services.data.ProcessService.addAllObjectsToIndex(ProcessService.java:343) ~[classes/:3.5.0]
        at org.kitodo.production.helper.IndexWorker.indexObjects(IndexWorker.java:98) ~[classes/:3.5.0]
        at org.kitodo.production.helper.IndexWorker.run(IndexWorker.java:69) ~[classes/:3.5.0]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
        at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:63) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:67) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.getResultSet(Loader.java:2322) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2075) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doQuery(Loader.java:956) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:357) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doList(Loader.java:2868) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doList(Loader.java:2850) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2682) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.list(Loader.java:2677) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.custom.CustomLoader.list(CustomLoader.java:338) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.internal.SessionImpl.listCustomQuery(SessionImpl.java:2186) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.internal.AbstractSharedSessionContract.list(AbstractSharedSessionContract.java:1204) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.query.internal.NativeQueryImpl.doList(NativeQueryImpl.java:177) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        ... 18 more
Caused by: java.sql.SQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'RECURSIVE process_children(id) as (        (SELECT id FROM process WHERE id = 28' at line 1
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120) ~[mysql-connector-j-8.0.31.jar:8.0.31]
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) ~[mysql-connector-j-8.0.31.jar:8.0.31]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:916) ~[mysql-connector-j-8.0.31.jar:8.0.31]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:972) ~[mysql-connector-j-8.0.31.jar:8.0.31]
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471) ~[c3p0-0.9.5.5.jar:0.9.5.5]
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:57) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.getResultSet(Loader.java:2322) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2075) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doQuery(Loader.java:956) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:357) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doList(Loader.java:2868) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.doList(Loader.java:2850) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2682) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.Loader.list(Loader.java:2677) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.loader.custom.CustomLoader.list(CustomLoader.java:338) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.internal.SessionImpl.listCustomQuery(SessionImpl.java:2186) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.internal.AbstractSharedSessionContract.list(AbstractSharedSessionContract.java:1204) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.query.internal.NativeQueryImpl.doList(NativeQueryImpl.java:177) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617) ~[hibernate-core-5.6.10.Final.jar:5.6.10.Final]
        ... 18 more
[ERROR] 2023-04-14 10:30:23.421 [pool-48-thread-1] IndexWorker - stop indexing after maximum amount of attempts
```

How to fix this?

**Comment by matthias-ronge** [comment]
Looks so. So I need to update MySQL?

**Comment by thomaslow** [comment]
Yes, upgrading to MySql >=8.0.1 would be the best long-term solution. Alternatively, you could build a war file that includes the pull request #5562. However the fix has a performance drawback, which would be a problem in case you have lots of parent processes with many children processes (e.g. newspaper processes).

**Comment by matthias-ronge** [comment]
FYI, we updated to MariaDB v. 10.1 (which is the version shipped with Ubuntu 18.04 “Bionic Beaver”) but that **does not** fix the problem. So, here too, a never version will be necessary.

**Comment by henning-gerhardt** [comment]
> FYI, we updated to MariaDB v. 10.1 (which is the version shipped with Ubuntu 18.04 “Bionic Beaver”) but that **does not** fix the problem. So, here too, a never version will be necessary.

You need at least MariaDB 10.3 or better 10.5. But I don't think that Ubuntu 18.04 are shipping them. Maybe include the MariaDB package repository if they provide a version for Ubuntu 18.04. Updating to Ubuntu 20.04 should maybe even work. Ubuntu 22.04 could be to new (maybe only with Java 17).

**Comment by thomaslow** [comment]
@matthias-ronge Please consider that Ubuntu 18.04 will stop to receive (free) security updates on 31 May 2023. So, upgrading Ubuntu would be a win-win.
